File: frontend/src/game/engine/applyMove.js
// Apply moves to state

import { PIECE_TYPES, PLAYERS, getOpponent, countPieces, cloneState } from './state';

/**
 * Apply a move to the state (returns new state, does not mutate)
 */
export function applyMove(state, move) {
  const newState = cloneState(state);
  const { from, to, captures, promotes } = move;
  const piece = newState.board[from.row][from.col];

  if (!piece) {
    console.warn('applyMove: No piece at source position', from);
    return state; // Return unchanged state instead of throwing
  }

  // Move the piece
  newState.board[from.row][from.col] = null;
  newState.board[to.row][to.col] = piece;

  // Remove captured pieces
  for (const capture of captures) {
    newState.board[capture.row][capture.col] = null;
    newState.capturedPieces[piece.player]++;
  }

  // Check for promotion
  const shouldPromote = promotes || (
    piece.type === PIECE_TYPES.MAN &&
    ((piece.player === PLAYERS.RED && to.row === 7) ||
     (piece.player === PLAYERS.WHITE && to.row === 0))
  );

  if (shouldPromote) {
    newState.board[to.row][to.col] = {
      ...piece,
      type: PIECE_TYPES.KING,
    };
  }

  // Record move in history
  newState.moveHistory.push({
    ...move,
    player: state.currentPlayer,
    promoted: shouldPromote,
    notation: generateNotation(move, piece, shouldPromote),
  });

  // Switch player
  newState.currentPlayer = getOpponent(state.currentPlayer);

  // Check for game over
  const opponentPieces = countPieces(newState, newState.currentPlayer);
  if (opponentPieces.total === 0) {
    newState.gameOver = true;
    newState.winner = state.currentPlayer;
  }

  return newState;
}

/**
 * Generate algebraic notation for a move
 */
function generateNotation(move, piece, promoted) {
  const fromSquare = String.fromCharCode(97 + move.from.col) + (8 - move.from.row);
  const toSquare = String.fromCharCode(97 + move.to.col) + (8 - move.to.row);
  const separator = move.isCapture ? 'x' : '-';
  const promotionSuffix = promoted ? '=K' : '';

  return `${fromSquare}${separator}${toSquare}${promotionSuffix}`;
}

/**
 * Undo the last move (returns new state)
 */
export function undoMove(state) {
  if (state.moveHistory.length === 0) {
    return state;
  }

  // For now, we'll implement a simple replay from start
  // A more sophisticated approach would store snapshots
  const newState = cloneState(state);
  const lastMove = newState.moveHistory.pop();

  // Handle both from and from_pos property names
  const from = lastMove.from || lastMove.from_pos;
  const to = lastMove.to;

  if (!from || !to) {
    console.warn('undoMove: Move missing from/to', lastMove);
    return state;
  }

  // Reverse the move
  const piece = newState.board[to.row][to.col];

  if (!piece) {
    console.warn('undoMove: No piece at destination', to);
    return state;
  }

  // If it was promoted, demote it
  if (lastMove.promoted) {
    piece.type = PIECE_TYPES.MAN;
  }

  // Move piece back
  newState.board[to.row][to.col] = null;
  newState.board[from.row][from.col] = piece;

  // Restore captured pieces
  const captures = Array.isArray(lastMove.captures) ? lastMove.captures : [];
  for (const capture of captures) {
    if (capture.piece) {
      newState.board[capture.row][capture.col] = capture.piece;
    } else {
      // Reconstruct piece from current player's opponent
      const opponent = getOpponent(lastMove.player);
      newState.board[capture.row][capture.col] = { player: opponent, type: PIECE_TYPES.MAN };
    }
    newState.capturedPieces[lastMove.player]--;
  }

  // Switch back to previous player
  newState.currentPlayer = lastMove.player;

  // Reset game over if applicable
  newState.gameOver = false;
  newState.winner = null;

  return newState;
}

/**
 * Check if current player has any legal moves
 */
export function hasLegalMoves(state) {
  const player = state.currentPlayer;
  const pieceCount = countPieces(state, player);

  if (pieceCount.total === 0) {
    return false;
  }

  // This is a simplified check - in practice we'd call generateLegalMoves
  // but we want to avoid circular dependency
  for (let row = 0; row < 8; row++) {
    for (let col = 0; col < 8; col++) {
      const piece = state.board[row][col];
      if (piece && piece.player === player) {
        // Check for any adjacent empty squares or jump opportunities
        const directions = piece.type === PIECE_TYPES.KING
          ? [[-1, -1], [-1, 1], [1, -1], [1, 1]]
          : piece.player === PLAYERS.RED
            ? [[1, -1], [1, 1]]
            : [[-1, -1], [-1, 1]];

        for (const [dRow, dCol] of directions) {
          const newRow = row + dRow;
          const newCol = col + dCol;

          // Check for regular move
          if (newRow >= 0 && newRow <= 7 && newCol >= 0 && newCol <= 7) {
            if (!state.board[newRow][newCol]) {
              return true;
            }

            // Check for jump
            const jumpRow = row + 2 * dRow;
            const jumpCol = col + 2 * dCol;
            if (jumpRow >= 0 && jumpRow <= 7 && jumpCol >= 0 && jumpCol <= 7) {
              const midPiece = state.board[newRow][newCol];
              if (midPiece && midPiece.player !== player && !state.board[jumpRow][jumpCol]) {
                return true;
              }
            }
          }
        }
      }
    }
  }

  return false;
}

/**
 * Get winner if game is over
 */
export function getWinner(state) {
  if (state.gameOver) {
    return state.winner;
  }

  // Check if current player has no moves
  if (!hasLegalMoves(state)) {
    return getOpponent(state.currentPlayer);
  }

  return null;
}

File: backend/app/game/apply_move.py
"""Apply moves to game state."""
from .state import GameState, Piece, Player, PieceType, clone_state, get_winner
from .moves import Move


def apply_move(state: GameState, move: Move) -> GameState:
    """Apply a move to the state (returns new state, does not mutate)."""
    new_state = clone_state(state)
    from_row, from_col = move.from_pos["row"], move.from_pos["col"]
    to_row, to_col = move.to["row"], move.to["col"]

    piece = new_state.board[from_row][from_col]

    if not piece:
        return state

    # Store original piece type before promotion
    original_type = piece.type

    # Move the piece
    new_state.board[from_row][from_col] = None
    new_state.board[to_row][to_col] = piece

    # Store captured pieces with their info for undo
    captured_pieces = []
    for capture in move.captures:
        cap_row, cap_col = capture["row"], capture["col"]
        captured_piece = new_state.board[cap_row][cap_col]
        if captured_piece:
            captured_pieces.append({
                "row": cap_row,
                "col": cap_col,
                "player": captured_piece.player.value,
                "type": captured_piece.type.value,
            })
        new_state.board[cap_row][cap_col] = None
        new_state.captured_pieces[piece.player.value] += 1

    # Promote if needed
    if move.promotes and piece.type == PieceType.MAN:
        piece.type = PieceType.KING

    # Record move in history with full info for undo
    move_record = {
        "from": {"row": from_row, "col": from_col},
        "to": {"row": to_row, "col": to_col},
        "captures": captured_pieces,
        "promotes": move.promotes,
        "player": piece.player.value,
        "original_type": original_type.value,  # Store for undo
    }
    new_state.move_history.append(move_record)

    # Switch player
    new_state.current_player = Player.WHITE if state.current_player == Player.RED else Player.RED

    # Check for game over
    winner = get_winner(new_state)
    if winner:
        new_state.game_over = True
        new_state.winner = winner

    return new_state


def undo_move(state: GameState) -> GameState:
    """Undo the last move (returns new state)."""
    if not state.move_history:
        return state

    new_state = clone_state(state)
    last_move = new_state.move_history.pop()

    from_pos = last_move["from"]
    to_pos = last_move["to"]
    player = Player(last_move["player"])

    # Get the piece at destination
    piece = new_state.board[to_pos["row"]][to_pos["col"]]

    if piece:
        # Demote if it was promoted
        if last_move.get("promotes"):
            piece.type = PieceType.MAN

        # Move piece back
        new_state.board[to_pos["row"]][to_pos["col"]] = None
        new_state.board[from_pos["row"]][from_pos["col"]] = piece

    # Restore captured pieces
    for capture in last_move.get("captures", []):
        captured_piece = Piece(
            player=Player(capture["player"]),
            type=PieceType(capture["type"]),
        )
        new_state.board[capture["row"]][capture["col"]] = captured_piece
        # Decrease captured count
        new_state.captured_pieces[player.value] -= 1

    # Switch back to previous player
    new_state.current_player = player

    # Reset game over
    new_state.game_over = False
    new_state.winner = None

    return new_state
